src/modes/ModeEstimation.jl

```diff
@@ -0,0 +1,318 @@
+
+Generate an unlinked (with no variable transformions) version of an existing `OptimLogDensity`.
+"""
+function Bijectors.invlink(f::OptimLogDensity)
+    init = VarInfo(f.model, f.context)
+    return OptimLogDensity(f.model, f.context, init, false)
+end
+
+"""
+    link(f::OptimLogDensity)
+
+Generate an linked (with variable transformions) version of an existing `OptimLogDensity`.
+"""
+function Bijectors.link(f::OptimLogDensity)
+    init = VarInfo(f.model, f.context)
+    DynamicPPL.link!(init, DynamicPPL.SampleFromPrior())
+    return OptimLogDensity(f.model, f.context, init, false)
+end
+
+"""
+    ModeResult{
+        V<:NamedArrays.NamedArray, 
+        M<:NamedArrays.NamedArray, 
+        O<:Optim.MultivariateOptimizationResults, 
+        S<:NamedArrays.NamedArray
+    }
+
+A wrapper struct to store various results from a MAP or MLE estimation.
+"""
+struct ModeResult{
+    V<:NamedArrays.NamedArray, 
+    O<:Optim.MultivariateOptimizationResults,
+    M<:OptimLogDensity
+} <: StatsBase.StatisticalModel
+    "A vector with the resulting point estimates."
+    values :: V
+    "The stored Optim.jl results."
+    optim_result :: O
+    "The final log likelihood or log joint, depending on whether `MAP` or `MLE` was run."
+    lp :: Float64
+    "The evaluation function used to calculate the output."
+    f :: M
+end
+#############################
+# Various StatsBase methods #
+#############################
+
+function Base.show(io::IO, m::ModeResult)
+    println(io, Printf.@sprintf("ModeResult with minimized lp of %.2f", m.lp), "\n")
+    show(io, m.values)
+end
+
+function StatsBase.coeftable(m::ModeResult)
+    # Get columns for coeftable.
+    terms = StatsBase.coefnames(m)
+    estimates = m.values.array[:,1]
+    stderrors = StatsBase.stderror(m)
+    tstats = estimates ./ stderrors
+
+    StatsBase.CoefTable([estimates, stderrors, tstats], ["estimate", "stderror", "tstat"], terms)
+end
+
+function StatsBase.informationmatrix(m::ModeResult; hessian_function=ForwardDiff.hessian, kwargs...)
+    # Calculate Hessian and information matrix.
+    varnames = StatsBase.coefnames(m)
+    f = invlink(m.f)
+    info = inv(hessian_function(x -> f(x), m.values.array[:, 1]))
+    return NamedArrays.NamedArray(info, (varnames, varnames))
+end
+
+StatsBase.coef(m::ModeResult) = m.values
+StatsBase.coefnames(m::ModeResult) = names(m.values)[1]
+StatsBase.params(m::ModeResult) = StatsBase.coefnames(m)
+StatsBase.vcov(m::ModeResult) = StatsBase.informationmatrix(m)
+StatsBase.loglikelihood(m::ModeResult) = m.lp
+
+####################
+# Optim.jl methods #
+####################
+
+"""
+    Optim.optimize(model::Model, ::MLE, args...; kwargs...)
+
+Compute a maximum likelihood estimate of the `model`.
+
+# Examples
+
+```julia
+@model function f(x)
+    m ~ Normal(0, 1)
+    x ~ Normal(m, 1)
+end
+
+model = f(1.5)
+mle = optimize(model, MLE())
+
+# Use a different optimizer
+mle = optimize(model, MLE(), NelderMeade())
```

Comment:
```suggestion
mle = optimize(model, MLE(), NelderMead())
```
I guess it would be nice to use a julia-repl code block here (see https://docs.julialang.org/en/latest/manual/documentation/, one could also write doctests).